plugins/wacom-i2c/fu-wac-device.c

```diff
@@ -0,0 +1,478 @@
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "expected firmware format is 'ihex', got '%s'",
+			     dfu_firmware_format_to_string (dfu_firmware_get_format (firmware)));
+		return FALSE;
+	}
+
+	/* use the correct image from the firmware */
+	image = dfu_firmware_get_image_default (firmware);
+	if (image == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "no firmware image");
+		return FALSE;
+	}
+	element = dfu_image_get_element_default (image);
+	if (element == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "no element in image");
+		return FALSE;
+	}
+	g_debug ("using element at addr 0x%0x",
+		 (guint) dfu_element_get_address (element));
+
+	/* check start address and size */
+	if (dfu_element_get_address (element) != priv->flash_base_addr) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "base addr invalid: 0x%05x",
+			     (guint) dfu_element_get_address (element));
+		return FALSE;
+	}
+	fw_new = dfu_element_get_contents (element);
+	if (g_bytes_get_size (fw_new) > priv->flash_size) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "size is invalid: 0x%05x",
+			     (guint) g_bytes_get_size (fw_new));
+		return FALSE;
+	}
+
+	/* we're in bootloader mode now */
+	if (!fu_wac_device_check_mode (self, error))
+		return FALSE;
+	if (!fu_wac_device_set_version_bootloader (self, error))
+		return FALSE;
+
+	/* flash chunks */
+	chunks = fu_chunk_array_new_from_bytes (fw_new, priv->flash_base_addr,
+						0x00,	/* page_sz */
+						priv->flash_block_size);
+	return klass->write_firmware (device, chunks, error);
+}
+
+gboolean
+fu_wac_device_set_feature (FuWacDevice *self, guint8 report_id,
+			   const guint8 *data, guint datasz, GError **error)
+{
+	FuWacDevicePrivate *priv = GET_PRIVATE (self);
+	gint ret = -1;
+	gint bufsz = SET_FEATURE_SIZE + datasz;
+	const guint16 cmdreg = FU_WAC_DEVICE_COMM_REG;
+	const guint16 datareg = FU_WAC_DEVICE_DATA_REG;
+	g_autofree guint8 *buf = g_malloc0 (bufsz);
+
+	buf[0] = (guint8)(cmdreg & 0x00ff);
+	buf[1] = (guint8)((cmdreg & 0xff00) >> 8);
+	buf[2] = (REPORT_TYPE_FEATURE << 4) | report_id;
+	buf[3] = CMD_SET_FEATURE;
+	buf[4] = (guint8)(datareg & 0x00ff);
+	buf[5] = (guint8)((datareg & 0xff00) >> 8);
+	if ((datasz + 2) > 255) {
+		buf[6] = (guint8)((datasz + 2) & 0x00ff);
+		buf[7] = (guint8)(((datasz + 2) & 0xff00) >> 8);
+	} else {
+		buf[6] = (guint8)(datasz + 2);
+		buf[7] = (guint8)(0x00);
+	}
+	memcpy (buf + SET_FEATURE_SIZE, data, datasz);
+	ret = write (priv->fd, buf, bufsz);
+	if (ret != bufsz) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "failed sent bytes: %d", ret);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/* use ioctl for using I2C restart method to communicate */
+gboolean
+fu_wac_device_get_feature (FuWacDevice *self, guint8 report_id,
```

Comment:
Does the device also register with hidraw?  I wonder if it would be "cleaner" to operate at that level instead if you can avoid needing to do the really _low_ level communication,

Reply:
That's a good question. I don't have hardware yet, we're just at the start of the process with Wacom Japan.

Reply:
I suspect presence of hidraw device is going to be mostly a function of how the ACPI tables are assembled.  If there is a proper ACPI entry that will register this as a HID device to the OS (which I think there should be; otherwise how would touch over HID-I2C work ?).

Reply:
> otherwise how would touch over HID-I2C work

That is a good point. I'll let you know when I get some answers.